Blog posts: a deep dive into a Go compiler bug fix

agarciamontoro opened the pull request:
#### Summary
This PR adds three blog posts (which were originally conceived as a single one) on how I investigated and fixed my first bug in the Go compiler. You can preview the three parts here:
- [Part 1](https://mattermost.github.io/mattermost-developer-documentation/803/blog/optimizing-go-compiler-1/)
- [Part 2](https://mattermost.github.io/mattermost-developer-documentation/803/blog/optimizing-go-compiler-2/)
- [Part 3](https://mattermost.github.io/mattermost-developer-documentation/803/blog/optimizing-go-compiler-3/)

The intended audience for the posts are developers that do not have any experience contributing to compilers.

This is a long long read, so feel free to review each post separately. Also, feel free to decline the review if you do not have time, no problem at all! And, finally, this is not high priority at all, so no rush with the review, I already took my time to write it, so we can take more time to review it.

@justinegeffen and @cwarnermm: I did not know who to ping here, so I pinged both of you, just so you can decide who has the bad luck of reviewing this.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-32531

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,227 @@
+community: alejandro.garcia
+---
+
+[bf48163e8f2b604f3b9e83951e331cd11edd8495](https://github.com/golang/go/commit/bf48163e8f2b604f3b9e83951e331cd11edd8495). That is the hash of one of the 47791 commits that, at the time of this writing, build the whole git history of the Go compiler. It is definitely not the most important commit there, and most of the other ones probably deserve way more attention. But I like `bf48163`. It is a g ...
```
> That is the hash of one of the 47791 commits that, at the time of this writing, build the whole git history of the Go compiler.

The last part reads a bit odd to me. I understand what you are trying to say, but what does "build the whole git history" mean? Isn't it just building the Go compiler?

0/5, but perhaps something like this is equivalent:

> [bf48163e8f2b604f3b9e83951e331cd11edd8495](https://github.com/golang/go/commit/bf48163e8f2b604f3b9e83951e331cd11edd8495). That is the hash of one of the 47791 commits of the Go programming language, at the time of this writing.

agarciamontoro replied:
> The last part reads a bit odd to me. I understand what you are trying to say, but what does "build the whole git history" mean? Isn't it just building the Go compiler?

My intention with this was to say that there are 47791 commits in the git history, each and everyone contributing to create the Go compiler. It is not build as in `make build`, but in the sense of constitute, of contributing to the history of the development of the compiler.

Would it be more clear to say something like this?

> That is the hash of one of the 47791 commits that, at the time of this writing, **form/populate/build up?** the whole git history of the Go compiler.

If it's still confusing, I'll use something like what you suggested.

agnivade replied:
Right, that's more or less what I thought. It's the start of the post, so I'll leave it to whatever you feel is the best. :)

agarciamontoro replied:
Resolving this conversation too to remove some noise from the PR. Thanks again!

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,227 @@
+-   [Part 2](/blog/optimizing-go-compiler-2): what the bug actually is.
+-   [Part 3](/blog/optimizing-go-compiler-3): how to fix the bug.
+
+Do not get discouraged for the length of the content: it does not correlate with the complexity of the issue, but with the level of detail I tried to use when explaining it. I wanted to make sure I explained the ins and outs of how I approached the fix, and I do not assume any knowledge about compilers from the reader, so I tried to discuss everything in chapter and verse.
```
There shouldn't be a period at the end of "explaining it" I think. Maybe place it at the end of "approached the fix". But I'll leave the grammatical stuff to Justine :P

agarciamontoro replied:
Hm, not sure. For context, this was the structure in my head:

Do not get discouraged for X: it is not because of Y, but because of Z. And Z happens because I had these intentions.

But yeah, I agree that Justine has the last word!

agnivade replied:
Ah ok, I read it slightly differently.

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
````diff
@@ -0,0 +1,227 @@
+}
+```
+
+It is now more clear what the function does. It stores the contents of the second argument, the `src` array, into the first argument, the slice `dest`. It uses the [encoding/binary](https://golang.org/pkg/encoding/binary/) package for reading and writing the 8 bytes as an unsigned integer of 64 bits. For doing the copy, we use a temporal variable, `temp`.
````
maybe "temporary" variable is better. "temporal" sounds like a sci-fi term :P

agarciamontoro replied:
Good catch, that's a mistake I always make!

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,227 @@
+
+Before diving into the bug, let's pause for a bit to learn something more about how the Go compiler works.
+
+For anyone who does not know, a compiler is a translator: it converts a piece of code in a language (in our case, Go) to an equivalent piece of code in another language (in our case, assembly code, whose specific instructions depend on the specific architecture we want to run the executable in). Hopefully, the meaning of what we wrote in Go (the behavior of our program), is maintained throughout  ...
```
You are already writing the text here, so I'd perhaps avoid saying "For anyone who does not know". It might make a beginner feel bad, that they "should" know what a compiler is. Perhaps better to just write "Essentially, a compiler is .."

agarciamontoro replied:
Completely true, thanks for the suggestion!

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
````diff
@@ -0,0 +1,227 @@
+
+We can compute the value of `a` by literally making the multiplication `5 times 16`. But it is known that multiplication instructions are quite CPU-expensive. What if we could optimize this line of code by avoiding the multiplication?
+
+It turns out we can, because `16` is a power of `2`. And multiplying by `2^n` is the same as shifting the number to the left `n` places. This is easy to see in binary, where 5 is:
+
+```
+00000101
+```
+
+If we multiply 5 by 2, we get 10, which in binary is:
+
+```
+00001010
+```
+
+And again by 2, we get 20:
+
+```
+00010100
+```
+
+See how we are shifting everything to the left each time we multiply by 2? Then, multiplying by 16 is the same as shifting to the left 4 times, as 16 is `2^4`:
+
+```
+00000101 // 5
+01010000 // 5*(2^4) = 80
+```
````
I feel like this level of detail is unnecessary. We can just skip this part with just a one-line explanation that doubling is the same as a single left shift.

agarciamontoro replied:
I agree. I erred on the side of over-explaining, and this is a good example. Changed, thanks!

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,227 @@
+
+## Connecting all together
+
+In order to investigate the issue and solve it, we will need to set up the development environment for the Go compiler. I encourage you to read the official [documentation](https://golang.org/doc/contribute): it is comprehensive and easy to understand. There are some particularities to the process, as the need to use Gerrit instead of Github, but it is really straightforward.
```
It is not really necessary to use Gerrit. You can just use GitHub too. https://github.com/golang/go/wiki/GerritBot#workflow

agarciamontoro replied:
That's true, there's no "need" to use it, but [it is encouraged](https://golang.org/doc/contribute#sending_a_change_gerrit) (although using Github for newcomers is [also encouraged](https://golang.org/doc/contribute#sending_a_change_github)). I left the phrase without the explanation between the commas. Thanks!

agnivade reviewed (commented):
Amazing piece of work! Just an initial pass on the first part.

justinegeffen commented:
I'll review this, thanks @agarciamontoro!

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
```diff
@@ -0,0 +1,343 @@
+	v153 (6) = CMPQconst <flags> [8] v9
+UGT v153 → b43 b3 (likely) (6)
+
+b3: ← b1
+
+	v12 (6) = LoweredPanicBoundsC <mem> [0] v8 v9 v1
+Exit v12 (6)
+
+b24:
+BlockInvalid (+83)
+
+b25:
+BlockInvalid (84)
+
+b27:
+BlockInvalid (85)
+
+b29:
+BlockInvalid (86)
+
+b31:
+BlockInvalid (87)
+
+b33:
+BlockInvalid (88)
+
+b35:
+BlockInvalid (89)
+
+b37:
+BlockInvalid (90)
+
+b39:
+BlockInvalid (91)
```
I am wondering if it's really necessary to mention this part. This is just the raw output, and not really helpful in the context of the post. It can be redacted to save some space.

agarciamontoro replied:
I removed all the invalid blocks, but left `b3`, and added a brief explanation on what it does. See https://github.com/mattermost/mattermost-developer-documentation/pull/803/commits/4c3766ee500569280e9b6282ac68aea9c3bc3e8d, thanks!

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
```diff
@@ -0,0 +1,343 @@
+
+The interesting blocks are `b1`, which initialize the values with the arguments of the function, and `b43`, which makes the actual copy.
+
+## Definition block
```
I'd suggest that this section on "Definition block" can be skipped. You are not really going line by line and explaining each and every aspect of everything. Just that this has the first and second argument. IMO - this can be condensed to just say that block `b1` just passes the params.

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
```diff
@@ -0,0 +1,343 @@
+func copyArrayToSlice(dest []byte, src *[8]byte) {
+	_ = dest[8]                                // 1. bounds check
+	temp := binary.LittleEndian.Uint64(src[:]) // 2. read the contents of the source array into temp
+	binary.LittleEndian.PutUint64(dest, temp)  // 3. write the contents into the destiny slice
```
destiny slice is a really cool slice. Let's fix it ;)

agarciamontoro replied:
Oh man, we use the same word in Spanish for both destiny and destination :joy: Fixed in https://github.com/mattermost/mattermost-developer-documentation/pull/803/commits/bcd613fd20785b2de630c486a25b70caa5dfc2fe, thanks!

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
````diff
@@ -0,0 +1,343 @@
+v248 (84) = Arg <*byte> {dest}
+```
+
+`v248` is the first argument of the function, `dest`, which is a pointer to bytes (we do not know the length here, as it was defined as a slice, which has a variable length).
````
It is not really a "pointer to bytes", it is a "slice header" - which is comprised of three things - the base address of the array, length, and capacity. That's what those three lines - v8, v9, v248 are I think. Anyways, I think it's perhaps simpler to just skip this section.

agarciamontoro replied:
I was going to remove the section per your previous comment, but then read this one and realized I had not understood this block completely. So, instead of removing the section, I now think I should actually expand it :blush:

But I need some help with this. Could you let me know if this makes sense:

```
v1 (?) = InitMem <mem>
```

Init memory, easy peasy.

```
v7 (5) = Arg <*[8]byte> {src} (b.ptr[*byte], src[*[8]byte])
```

This represents the `src` array. The `<*[8]byte> is the type. I *guess* `{src}` is the GCnode. But I'm not sure what `(b.ptr[*byte], src[*[8]byte])` is, actually. I don't even know what that `b` is.

```
v8 (?) = MOVQconst <int> [8] (b.cap[int], b.len[int])
```

I guess this has to do with the length of the slice. Are we simply storing an `8` into the slice capacity and length?

```
v9 (+6) = Arg <int> {dest} [8] (b.len[int], dest+8[int])
```

Again, not sure what this argument is. We have three `Arg` lines, but the function only gets two arguments. I think this has to do with the bounds check in the first line of the function, because `v9` is then used in the `CMPQconst` line to decide which block we're going to next, but I'm not sure what the exact meaning of this line is.

```
v248 (84) = Arg <*byte> {dest}
```

This is a pointer to where the *data* of the `dest` slice lives, right?

```
v153 (6) = CMPQconst <flags> [8] v9
```

Finally, we're comparing `v9` to 8 (?) Is `v9` the length of the slice, maybe?

agnivade replied:
I can probably guess some parts of it, but I am nowhere near an SME on this matter. Let me ping @josharian to lend some of his expertise on this.

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
```diff
@@ -0,0 +1,343 @@
+
+This first line of the block, then, seems perfectly fine: it loads the whole contents of the `src` array into memory (which is effectively the `temp` variable we defined in the code), and it does it with a single instruction. We can no longer optimize this.
+
+But remember that the original function did two things: first, it loaded the contents of the `src` array into a temporal variable, and then it stored those contents into the `dest` slice. We have already loaded the contents into memory with the line we discussed above, so the rest of the block should do the rest of the work: store those contents from memory into the `dest` slice (which was repres ...
```
temporal

agarciamontoro replied:
Same thing as with destiny/destination. We reuse a lot of words in Spanish, I guess, and that shows :joy: Fixed in https://github.com/mattermost/mattermost-developer-documentation/pull/803/commits/bcd613fd20785b2de630c486a25b70caa5dfc2fe, thanks!

agnivade commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+
+With the addition of this rule, the test passes, so we solved the issue!
+
+...or didn't we? Sorry, there's just one more wee thing to do. In the issue comments, [@mundaym](https://github.com/mundaym) pointed out [an important thing](https://github.com/golang/go/issues/41663#issuecomment-699904531):
```
Should this be "or did we?"

agarciamontoro replied:
Fixed in https://github.com/mattermost/mattermost-developer-documentation/pull/803/commits/1c7164da0e23ad3fe203044b55e41bd39a29e1c7, thanks!

reflog reviewed (approved):
Don't have much to add on top of Agniva's comments, the article is awesome, I definitely learned a lot. Very well done (both the patch and the blog post)

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,209 @@
+
+## Connecting all together
+
+In order to investigate the issue and solve it, we will need to set up the development environment for the Go compiler. I encourage you to read the official [documentation](https://golang.org/doc/contribute): it is comprehensive and easy to understand. There are some particularities to the process, but it is really straightforward.
```
```suggestion
In order to investigate the issue and solve it, we'll need to set up the development environment for the Go compiler. I encourage you to read the official [documentation](https://golang.org/doc/contribute): it's comprehensive and easy to understand. There are some particularities to the process, but it's generally straightforward.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,209 @@
+
+In order to investigate the issue and solve it, we will need to set up the development environment for the Go compiler. I encourage you to read the official [documentation](https://golang.org/doc/contribute): it is comprehensive and easy to understand. There are some particularities to the process, but it is really straightforward.
+
+For the specifics we will need to use in these posts, we can summarize a couple of things. All the commands in this section assume that you have cloned the Go repository and have a shell open in its root directory.
```
```suggestion
For the specifics we'll need to use in these posts, we can summarize a couple of things. All the commands in this section assume that you've cloned the Go repository and have a shell open in its root directory.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,209 @@
+
+### How to apply the changes done to the rewrite rules
+
+We already know what the rewrite rules are, but not how they are applied. Although the rules live in a series of `.rules` files, the actual code used to match and apply them is normal Go code. With one peculiarity: it is automatically generated from the `.rules` files.
```
```suggestion
We already know what the rewrite rules are, but not how they're applied. Although the rules live in a series of `.rules` files, the actual code used to match and apply them is normal Go code. With one peculiarity: it's automatically generated from the `.rules` files.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,209 @@
+
+For example, the rules for AMD64 are declared in the file `src/cmd/compile/internal/ssa/gen/AMD64.rules `, but the automatically generated code that applies those rules live in `src/cmd/compile/internal/ssa/rewriteAMD64.go`.
+
+So: whenever we manually update a rewrite rule, we need to re-generate the corresponding Go code. Doing so it's trivial, and it comes down to this:
```
```suggestion
So: whenever we manually update a rewrite rule, we need to re-generate the corresponding Go code. Doing so is trivial, and comes down to this:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,209 @@
+
+### How to compile the compiler and test it
+
+And now for the fun part: how to compile the compiler. Once we have generated the automatic code, we want to build the compiler from source with the new changes. Doing it is also as trivial as calling a script:
```
```suggestion
And now for the fun part: how to compile the compiler. Once we've generated the automatic code, we want to build the compiler from source with the new changes. Doing it is as trivial as calling a script:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,209 @@
+
+If everything was correctly setup, you should see your local Go directory when doing `which go`.
+
+And last, but not least, when you have applied your changes, you will want to know that you did not break anything else, so you should run the whole test suite at least once before you push your changes! Doing so is again one script call away:
```
```suggestion
And last, but not least, when you've applied your changes, you'll want to know that you didn't break anything else, so you should run the whole test suite at least once before you push your changes! Doing so is again one script call away:
```

justinegeffen reviewed (commented):
This is great, @agarciamontoro! I have part two and three to review still but part one is looking good to me. :) Just some minor grammatical suggestions.

agarciamontoro commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,209 @@
+
+Not only that, it is my first contribution _ever_ to any compiler!
+
+What's really interesting here is that I am not an expert in Go---nor in compilers in general---at all. Sure, I know my way through the language to write software using it, but I had never thought I could be able to modify the compiler itself.
```
Just a sanity check: the three dashes render like this:
![image](https://user-images.githubusercontent.com/3924815/119796834-002c8e00-beda-11eb-9457-f027fb8911f6.png)

While only one dash renders as follows:
![image](https://user-images.githubusercontent.com/3924815/119796762-ee4aeb00-bed9-11eb-99ac-2e615d5d7895.png)

Is this what we want, @justinegeffen?

justinegeffen replied:
Great question! So they're both correct imo. I'm 0/5 as to which we use so let's go with the original. :)

agarciamontoro commented on site/content/blog/2021-04-23-optimizing-go-compiler-1.md:
```diff
@@ -0,0 +1,209 @@
+
+Essentially, a compiler is a translator: it converts a piece of code in a language (in our case, Go) to an equivalent piece of code in another language (in our case, assembly code, whose specific instructions depend on the specific architecture we want to run the executable in). Hopefully, the meaning of what we wrote in Go (the behavior of our program), is maintained throughout the process of tr ...
+
+In the Go compiler, this complex process is divided into phases. There are _a lot_ of them, and we will not discuss the structure in this blog post---see [`cmd/compile/README.md`](https://github.com/golang/go/blob/0c93b16d015663a60ac77900ca0dcfab92310790/src/cmd/compile/README.md) for a high-level view of the phases, and take a look directly [at the code](https://github.com/golang/go/blob/0c93b16 ...
```
@justinegeffen, same question as before.

agarciamontoro commented:
Thank you for the awesome review, @justinegeffen! I added all of your suggestions, except for the two changing the dashes, where I left a comment.

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
```diff
@@ -0,0 +1,320 @@
+community: alejandro.garcia
+---
+
+This is the second part of a series of posts on the compiler of Go. Make sure you have read [the first part](/blog/optimizing-go-compiler-1) before going on.
```
```suggestion
This is the second part of a series of posts on the Go compiler. Make sure you've read [the first part](/blog/optimizing-go-compiler-1) before going on.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
````diff
@@ -0,0 +1,320 @@
+}
+```
+
+The data in this struct defines what the `MOVQload` instruction does; the corresponding code in assembly, `MOVQ`; its returning type, `UInt64`; or the number of arguments it receives, 2. The comment does a pretty good job explaining what the instruction does: it loads 8 bytes from the data pointed to by the first argument (plus some auxiliary arguments, if present, which we will see are constant  ...
````
```suggestion
The data in this struct defines what the `MOVQload` instruction does; the corresponding code in assembly, `MOVQ`; its returning type, `UInt64`; or the number of arguments it receives, 2. The comment does a pretty good job explaining what the instruction does: it loads 8 bytes from the data pointed to by the first argument (plus some auxiliary arguments, if present, which we'll see are constant numbers in square brackets) into the second, which should represent the memory. If we go back to our line:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
````diff
@@ -0,0 +1,320 @@
+    one as temp                ┗ src
+```
+
+we now can understand that this loads the contents from the `src` argument (remember that `v7`represented`src`), into the memory we initialized at the very beginning (represented by `v1`). There are no auxiliary arguments, so we can safely forget about those! The result of this instruction is represented by the `v34` value, which we can understand as the `temp` variable.
````
```suggestion
We now can understand that this loads the contents from the `src` argument (remember that `v7`represented`src`), into the memory we initialized at the very beginning (represented by `v1`). There are no auxiliary arguments, so we can safely forget about those! The result of this instruction is represented by the `v34` value, which we can understand as the `temp` variable.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
```diff
@@ -0,0 +1,320 @@
+
+This first line of the block, then, seems perfectly fine: it loads the whole contents of the `src` array into memory (which is effectively the `temp` variable we defined in the code), and it does it with a single instruction. We can no longer optimize this.
+
+But remember that the original function did two things: first, it loaded the contents of the `src` array into a temporary variable, and then it stored those contents into the `dest` slice. We have already loaded the contents into memory with the line we discussed above, so the rest of the block should do the rest of the work: store those contents from memory into the `dest` slice (which was repre ...
```
```suggestion
But remember that the original function did two things: First, it loaded the contents of the `src` array into a temporary variable, and then it stored those contents into the `dest` slice. We've already loaded the contents into memory with the line we discussed above, so the rest of the block should do the rest of the work: Store those contents from memory into the `dest` slice (which was represented by value `v248`). Let's see the rest of the block again:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
````diff
@@ -0,0 +1,320 @@
+v255 (91) = MOVBstore <mem> [7] v248 v180 v243
+```
+
+See that the values defined in the second, third and fourth lines are used in the three last lines. Let's rewrite the block, simply replacing the values where they are used:
````
```suggestion
See that the values defined in the second, third, and fourth lines are used in the three last lines. Let's rewrite the block, simply replacing the values where they're used:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
````diff
@@ -0,0 +1,320 @@
+v255 (91) = MOVBstore <mem> [7] v248 (SHRQconst <uint64> [56] v34) v243
+```
+
+SPOILER: these are the lines that store to the slice what we just read into memory. As you can see, we are using four instructions instead of one, which is what we are going to optimize!
````
```suggestion
SPOILER: These are the lines that store to the slice what we just read into memory. As you can see, we're using four instructions instead of one, which is what we're going to optimize!
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
```diff
@@ -0,0 +1,320 @@
+
+SPOILER: these are the lines that store to the slice what we just read into memory. As you can see, we are using four instructions instead of one, which is what we are going to optimize!
+
+But let's not get ahead of ourselves! To better understand what is going on here, we need to understand what `MOVBstore`, `MOVLstore` and `MOVWstore` do, as well as `SHRQconst`. The first set of instructions look quite similar to the one we already understand, `MOVQload`.
```
```suggestion
But let's not get ahead of ourselves! To better understand what's going on here, we need to understand what `MOVBstore`, `MOVLstore`, and `MOVWstore` do, as well as `SHRQconst`. The first set of instructions look quite similar to the one we already understand: `MOVQload`.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
````diff
@@ -0,0 +1,320 @@
+}
+```
+
+Again, the comment does a good job explaining what is going on. The instruction receives three arguments, and it stores a single byte from the memory pointed to by the first argument (plus the auxiliary arguments) into the second argument. The state of the memory is represented by the third argument. The `typ` of the struct tells us that what this instruction returns is the new state of the memor ...
````
```suggestion
Again, the comment does a good job explaining what's going on. The instruction receives three arguments, and it stores a single byte from the memory pointed to by the first argument (plus the auxiliary arguments) into the second argument. The state of the memory is represented by the third argument. The `typ` of the struct tells us that what this instruction returns is the new state of the memory after the operation. In short, this instruction gets a single byte from a place in memory and stores it somewhere else.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
````diff
@@ -0,0 +1,320 @@
+}
+```
+
+It's a good ol' shift to the right! That is: an operation that takes a binary number, and shifts it to the right a specific amount of places, specified by the `auxint` argument (the constant integer we see in square brackets). If shifting to the left was the same as multiplying by a power of two, shifting a number to the right `n` places means halving it `n` times, which is equal to dividing it b ...
````
```suggestion
It's a good ol' shift to the right! That is: An operation that takes a binary number, and shifts it to the right a specific number of places, specified by the `auxint` argument (the constant integer we see in square brackets). If shifting to the left was the same as multiplying by a power of two, shifting a number to the right `n` places means halving it `n` times, which is equal to dividing it between `2^n`.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
```diff
@@ -0,0 +1,320 @@
+
+We have all the ingredients now:
+
+-   `MOVXstore` stores the contents from the second argument into the first one, with the amount of bytes moved depending on the prefix `X`
```
```suggestion
-   `MOVXstore` stores the contents from the second argument into the first one, with the number of bytes moved depending on the prefix `X`.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
```diff
@@ -0,0 +1,320 @@
+We have all the ingredients now:
+
+-   `MOVXstore` stores the contents from the second argument into the first one, with the amount of bytes moved depending on the prefix `X`
+-   `SHRQconst` shifts the number in its only argument a number of places specified by the auxiliary argument, the number in square brackets
```
```suggestion
-   `SHRQconst` shifts the number in its only argument a number of places specified by the auxiliary argument, the number in square brackets.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
````diff
@@ -0,0 +1,320 @@
+         ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛      ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
+```
+
+Now we're storing a **L**ong word (4 bytes) from the second argument, `(SHRQconst <uint64> [8] v34)`, into the memory pointed to by `v248`, which is `dest`. But we have the `[1]` auxiliary argument! So we don't write directly into the memory pointed to by `v248`, but one byte after that pointer: that is, the first byte of `dest` is left untouched, and we write contents from the second byte on. In ...
````
```suggestion
Now we're storing a **L**ong word (4 bytes) from the second argument, `(SHRQconst <uint64> [8] v34)`, into the memory pointed to by `v248`, which is `dest`. But we have the `[1]` auxiliary argument! So we don't write directly into the memory pointed to by `v248`, but one byte after that pointer: That is, the first byte of `dest` is left untouched, and we write contents from the second byte on. In terms of Go code, as we're storing 4 bytes, it means that we're writing to `dest[1]`, `dest[2]`, `dest[3]` and `dest[4]`.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
```diff
@@ -0,0 +1,320 @@
+
+Now we're storing a **L**ong word (4 bytes) from the second argument, `(SHRQconst <uint64> [8] v34)`, into the memory pointed to by `v248`, which is `dest`. But we have the `[1]` auxiliary argument! So we don't write directly into the memory pointed to by `v248`, but one byte after that pointer: that is, the first byte of `dest` is left untouched, and we write contents from the second byte on. In ...
+
+But what 4 bytes do we write? That's what `(SHRQconst <uint64> [8] v34)` is telling us: we already wrote `temp[0]` in `dest[0]`, so it would make sense to read now from `temp[1]` on. And that's exactly what we are doing with the shift: as the auxiliary argument is `[8]`, that means that we are shifting the content of `v34` (or `temp`) to the right 8 bits. Those 8 bits are already in `dest[0]`, so ...
```
```suggestion
But what 4 bytes do we write? That's what `(SHRQconst <uint64> [8] v34)` is telling us: We already wrote `temp[0]` in `dest[0]`, so it would make sense to read now from `temp[1]` on. And that's exactly what we're doing with the shift: As the auxiliary argument is `[8]`, that means that we're shifting the content of `v34` (or `temp`) to the right 8 bits. Those 8 bits are already in `dest[0]`, so we can safely discard them and only read the next ones: from `temp[1]` on. As `MOVLstore` is writing 4 bytes, that means that we're getting 4 bytes' worth of data; that is: `temp[1]`, `temp[2]`, `temp[3]`, and `temp[4]`.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
```diff
@@ -0,0 +1,320 @@
+
+But what 4 bytes do we write? That's what `(SHRQconst <uint64> [8] v34)` is telling us: we already wrote `temp[0]` in `dest[0]`, so it would make sense to read now from `temp[1]` on. And that's exactly what we are doing with the shift: as the auxiliary argument is `[8]`, that means that we are shifting the content of `v34` (or `temp`) to the right 8 bits. Those 8 bits are already in `dest[0]`, so ...
+
+So that's it! That long, complex, full of arguments and auxiliary arguments instruction is simply doing this:
```
```suggestion
So that's it! That long, complex, full of arguments and auxiliary arguments, instruction is simply doing this:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
````diff
@@ -0,0 +1,320 @@
+v243 (90) = MOVWstore <mem> [5] v248 (SHRQconst <uint64> [40] v34) v219
+```
+
+This one stores 2 bytes (see the `W` prefix there?) into the memory pointed to by `v248` plus `[5]` bytes; that is, we write into `dest[5]` and `dest[6]`. And what we write is the content of `v34` (`temp`) shifted to the right `[40]` bits (or 5 bytes, as we have already copied from `temp[0]` to `temp[4]`). That's right, this line is doing the following:
````
```suggestion
This one stores 2 bytes (see the `W` prefix there?) into the memory pointed to by `v248` plus `[5]` bytes; that is, we write into `dest[5]` and `dest[6]`. And what we write is the content of `v34` (`temp`) shifted to the right `[40]` bits (or 5 bytes, as we've already copied from `temp[0]` to `temp[4]`). That's right, this line is doing the following:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
````diff
@@ -0,0 +1,320 @@
+v255 (91) = MOVBstore <mem> [7] v248 (SHRQconst <uint64> [56] v34) v243
+```
+
+Yup, we're simply copying the last byte from `temp` (skipping the first `[56]` bits, or 7 bytes!), into the `dest` slice (starting to write after `[7]` bytes, or 56 bits!). That's it, in terms of Go code, we are simply doing this:
````
```suggestion
Yup, we're simply copying the last byte from `temp` (skipping the first `[56]` bits, or 7 bytes!), into the `dest` slice (starting to write after `[7]` bytes, or 56 bits!). That's it, in terms of Go code, we're simply doing this:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-2.md:
````diff
@@ -0,0 +1,320 @@
+dst[7] = temp[7]
+```
+
+Wrapping up everything we've learned up until now, we are doing the following:
````
```suggestion
Wrapping up everything we've learned up until now, we're doing the following:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+title: "Fixing a bug in the Go compiler as a newbie: a deep dive (III)"
+heading: "Fixing a bug in the Go compiler as a newbie: a deep dive (III)"
+description: ""
+summary: "Finally, the last part of the series! We have learned how the compiler uses Static Single Assignment and the rewrite rules to write and optimize intermediate code, and we have dissected each and every line of the SSA generated by the compiler in the buggy function. Now it is time to use all that knowledge to apply a fix, which will come in the form of the most simple rewrite rule."
```
```suggestion
summary: "Finally, the last part of the series! We've learned how the compiler uses Static Single Assignment and the rewrite rules to write and optimize intermediate code, and we've dissected each and every line of the SSA generated by the compiler in the buggy function. Now it's time to use all that knowledge to apply a fix, which will come in the form of the most simple rewrite rule."
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+community: alejandro.garcia
+---
+
+This is the third and last part of a series of posts on the compiler of Go. Make sure you have read [the first](/blog/optimizing-go-compiler-1) and [the second](/blog/optimizing-go-compiler-2) parts before going on.
```
```suggestion
This is the third and last part of a series of posts on the Go compiler. Make sure you've read [the first](/blog/optimizing-go-compiler-1) and [the second](/blog/optimizing-go-compiler-2) parts before continuing.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+1. Trying to understand how the rules in the [generic.rules](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/gen/generic.rules) and [AMD64.rules](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/gen/AMD64.rules) files are applied during the compilation, and trying to fix the conflict tha ...
+2. Adding a new rule that explicitly matches that pattern and changes it with what we need.
+
+My first instinct was to go with 1: it is a more elegant approach, and would let us actually fix the issue, instead of patching it. It would also avoid the overhead of adding a new rule, which should be always checked and only applied when this specific, weird pattern happens.
```
```suggestion
My first instinct was to go with 1: It's a more elegant approach, and would let us actually fix the issue, instead of patching it. It would also avoid the overhead of adding a new rule, which should be always checked and only applied when this specific, weird pattern happens.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
````diff
@@ -0,0 +1,234 @@
+MOVQstore <mem> v248 v34 v1
+```
+
+With the actual pattern that we are seeing and the expected instruction that we want to see, we have all the pieces we need to build a new rule. In its simplest form, the rule has the form of `(actual pattern) => (expected pattern)`. In our case, it looks as follows:
````
```suggestion
With the actual pattern that we're seeing and the expected instruction that we want to see, we have all the pieces we need to build a new rule. In its simplest form, the rule has the form of `(actual pattern) => (expected pattern)`. In our case, it looks as follows:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
````diff
@@ -0,0 +1,234 @@
+  => (MOVQstore p1 w mem)
+```
+
+Comparing this with the actual pattern we described before, we can see a couple of changes: we have replaced `v248` with a variable named `p1`, `v34` with a variable named `w`, and `v1` with a variable named `mem` (as it represents the memory). With this renaming, we generalize the possible values those variables can have, but restrict the pattern to be matched only when the same value appears in ...
````
```suggestion
Comparing this with the actual pattern we described before, we can see a couple of changes: We've replaced `v248` with a variable named `p1`, `v34` with a variable named `w`, and `v1` with a variable named `mem` (as it represents the memory). With this renaming, we generalize the possible values those variables can have, but restrict the pattern to be matched only when the same value appears in every instance of the same variable. We've also removed the type specifications from the instructions, as they aren't important in the matching pattern.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
````diff
@@ -0,0 +1,234 @@
+  => (MOVQstore p1 w mem)
+```
+
+The only addition here is the `x1`, `x2` and `x3` labels, which name each of the values that correspond to the inner `MOV` instructions, and the boolean condition we have at the end of the matching pattern, which is actually literal Go code:
````
```suggestion
The only addition here are the `x1`, `x2`, and `x3` labels, which name each of the values that correspond to the inner `MOV` instructions, and the boolean condition we have at the end of the matching pattern, which is actually literal Go code:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
````diff
@@ -0,0 +1,234 @@
+  && clobber(x1, x2, x3)
+```
+
+This boolean condition is simply an additional restriction to the pattern matching. The rules specification say that this pattern will only match if the pattern is matched **and** if this condition is true.
````
```suggestion
This boolean condition is simply an additional restriction to the pattern matching. The rules specification says that this pattern will only match if the pattern is matched **and** if this condition is true.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+-   The [`Uses` field](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/value.go#L51-L52) is a field that we can access in every SSA value and, per its definition, is a counter of the number of times that specific value is used.
+-   The [`clobber` function](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/rewrite.go#L935-L945) both checks that the value is never used again and decrements the `.Uses` counter.
+
+In short, what we are checking with the boolean condition is that each of the inner `MOV` instructions are used only once, and making sure that they are never used again. This is a simple strategy to restrict the pattern matching to the specific case we want to fix.
```
```suggestion
In short, what we're checking with the boolean condition is that each of the inner `MOV` instructions are used only once, and making sure that they are never used again. This is a simple strategy to restrict the pattern matching to the specific case we want to fix.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+
+With the addition of this rule, the test passes, so we solved the issue!
+
+...or did we? Sorry, there's just one more wee thing to do. In the issue comments, [@mundaym](https://github.com/mundaym) pointed out [an important thing](https://github.com/golang/go/issues/41663#issuecomment-699904531):
```
```suggestion
...or did we? Sorry, there's just one more tiny thing to do. In the issue comments, [@mundaym](https://github.com/mundaym) pointed out [an important thing](https://github.com/golang/go/issues/41663#issuecomment-699904531):
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+
+> arm64, ppc64le and s390x all also do unaligned load/store merging.
+
+According to this, if we compile for any of those architectures, using the environment variable `GOARCH` with the values `arm64`, `ppc64le` or `s390x`, we may see a similar pattern, where the load is done with one instruction but the store is done in multiple pieces.
```
```suggestion
According to this, if we compile for any of those architectures, using the environment variable `GOARCH` with the values `arm64`, `ppc64le`, or `s390x`, we may see a similar pattern, where the load is done with one instruction but the store is done in multiple pieces.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+
+As you can see, the regex for each of the architectures is different, as the specific instructions differ for each of them.
+
+As I noted before, the only architecture that was not passing the test was `s390x`, so a new rule had to be added! The strategy to do so was exactly the same: study the pattern of instructions generated by the compiler and craft a rule to specifically match that pattern. [The new rule for S390X](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ss ...
```
```suggestion
As I noted before, the only architecture that was not passing the test was `s390x`, so a new rule had to be added! The strategy to do so was exactly the same: Study the pattern of instructions generated by the compiler and craft a rule to specifically match that pattern. [The new rule for S390X](https://github.com/golang/go/blob/bf48163e8f2b604f3b9e83951e331cd11edd8495/src/cmd/compile/internal/ssa/gen/S390X.rules#L1423-L1431) ended up taking this shape:
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
````diff
@@ -0,0 +1,234 @@
+  => (MOVDBRstore p1 w mem)
+```
+
+The structure is identical to the rule we studied in this series of posts, but if you are interested in the details of the S390X-specific instructions, feel free to dive into [the S390Xops.go file](https://github.com/golang/go/blob/master/src/cmd/compile/internal/ssa/gen/S390XOps.go) to learn what those instructions do.
````
```suggestion
The structure is identical to the rule we studied in this series of posts, but if you're interested in the details of the S390X-specific instructions, feel free to dive into [the S390Xops.go file](https://github.com/golang/go/blob/master/src/cmd/compile/internal/ssa/gen/S390XOps.go) to learn what those instructions do.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+
+That's it, we did it! This was a long journey to explain a single bug with a 30-line-long fix, but I hope it was an interesting one. At the very least, we learned about:
+
+-   Static Single Assignment (SSA), and how it makes code rewriting and optimizations easier to apply
```
```suggestion
-   Static Single Assignment (SSA), and how it makes code rewriting and optimizations easier to apply.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+That's it, we did it! This was a long journey to explain a single bug with a 30-line-long fix, but I hope it was an interesting one. At the very least, we learned about:
+
+-   Static Single Assignment (SSA), and how it makes code rewriting and optimizations easier to apply
+-   Rewrite rules, the tool to actually rewrite the code and apply those optimizations
```
```suggestion
-   Rewrite rules, the tool to actually rewrite the code and apply those optimizations.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+
+-   Static Single Assignment (SSA), and how it makes code rewriting and optimizations easier to apply
+-   Rewrite rules, the tool to actually rewrite the code and apply those optimizations
+-   A little bit of assembly, which made it possible to understand the bug and fix it
```
```suggestion
-   A little bit of assembly, which made it possible to understand the bug and fix it.
```

justinegeffen commented on site/content/blog/2021-04-23-optimizing-go-compiler-3.md:
```diff
@@ -0,0 +1,234 @@
+
+Also, the journey made us unveil some particularities of how the Go compiler works and the structure of its code, so I hope you are now even more ready to start contributing!
+
+If you are interested in becoming a Go contributor, go ahead, check the [issue tracker](https://github.com/golang/go/issues?q=is%3Aissue+is%3Aopen+label%3ANeedsFix) and comment on the issue you want to work on. There's no better way to learn how to contribute to a compiler than, well, contributing to a compiler.
```
```suggestion
If you're interested in becoming a Go contributor, go ahead, check the [issue tracker](https://github.com/golang/go/issues?q=is%3Aissue+is%3Aopen+label%3ANeedsFix) and comment on the issue you want to work on. There's no better way to learn how to contribute to a compiler than, well, contributing to a compiler.
```

justinegeffen reviewed (approved):
Thanks for this, @agarciamontoro! My feedback is only grammar and one type. Other than that this is awesome, well done! I'm approving to ensure you can move forward. :)

agarciamontoro commented:
Thank you for the huge review, Justine, I really appreciate it!

agarciamontoro commented:
This is on review by the marketing team in a separate Google Doc. I'll close this PR when the post is published :)

agarciamontoro commented:
This is already in the hands of the Marketing team and they'll finish reviewing it. So first reviewers of these posts: no worries, you don't need to review it again.

The changes here are structural. I converted the three posts into two stories: a study on SSA and rewrite rules, and a two-part investigation of the specific issue to fix. As the SSA post was a bit dull, I added a deeper explanation on one specific rewrite rule, which I had removed in a first draft because it made everything too long. With the restructure, it made sense to bring it back, and now I think both stories are more self-consistent. That's the only change in content, the other changes are just moving stuff around to adapt to this structure of two stories.

@BenLloydPearson, feel free to take the contents of this and move it to the Google Docs for your team to make the final review.

agarciamontoro commented:
Only the first post will be published. Closing this since it is already in the content team's hands. Thank you everyone who reviewed the whole thing, I will publish the whole original thing in my personal blog :)
